src/Backups/RestoreCoordinationDistributed.cpp

```diff
@@ -0,0 +1,148 @@
+#include <Backups/RestoreCoordinationDistributed.h>
+#include <Common/ZooKeeper/KeeperException.h>
+#include <Common/escapeForFileName.h>
+
+
+namespace DB
+{
+
+RestoreCoordinationDistributed::RestoreCoordinationDistributed(const String & zookeeper_path_, zkutil::GetZooKeeper get_zookeeper_)
+    : zookeeper_path(zookeeper_path_), get_zookeeper(get_zookeeper_)
+{
+    createRootNodes();
+}
+
+RestoreCoordinationDistributed::~RestoreCoordinationDistributed() = default;
+
+void RestoreCoordinationDistributed::createRootNodes()
+{
+    auto zookeeper = get_zookeeper();
+    zookeeper->createAncestors(zookeeper_path);
+    zookeeper->createIfNotExists(zookeeper_path, "");
+    zookeeper->createIfNotExists(zookeeper_path + "/paths_in_backup", "");
+    zookeeper->createIfNotExists(zookeeper_path + "/acquired", "");
+}
+
+void RestoreCoordinationDistributed::removeAllNodes()
+{
+    auto zookeeper = get_zookeeper();
+    zookeeper->removeRecursive(zookeeper_path);
+}
+
+void RestoreCoordinationDistributed::setOrGetPathInBackupForZkPath(const String & zk_path_, String & path_in_backup_)
```

Comment:
Unclear semantics of the function argument `path_in_backup_`. Looks like it can have a non-empty value (line 46) but can be overwritten? Maybe it is possible to return it as result?